### PR TITLE
Wait on progress

### PR DESCRIPTION
**Pull request by pb-cdunn**
Jim said there might be a problem of jobs being killed somehow on the Pod, so we wait forever on them.

I don't know quite what is going on, but I've made changes to let us detect whether a job is still making progress. The idea is simple:

* Every batch job will include a background process which writes to a progress-file periodically (every 60sec).
* The falcon job thread (i.e. not pypeflow) will `stat` that progress-file (mtime) while it waits for the done/exit files. If the progress-file does not change in 10min (arbitrary number), then we consider the job finished. (Success or failure is determined elsewhere.)
* Since we do not know when a queued job has begun, we first block on the *existence* of the progress-file. (We could timeout on that someday too, but that's a separate issue.)

It works fine in my testing.

The architecture is not great. I would rather put this functionality into a separate job-distribution task, which I plan to write soon. But Jim wanted something right away because of these mysterious problems on the Pod. It's easy to revert if we have a problem.

### CODE REVIEW

File: src/py/mains/run.py [hunk 1]
```diff
@@ -111,31 +120,141 @@ def wait_for_file(filename, task, job_name = ""):
...
+    wait_time = .2
+    while 1:
+        time.sleep(wait_time)
+        wait_time = 10 if wait_time > 10 else wait_time + .2
+        if os.path.exists(fn):
+            break
+
+def wait_for_file_with_progress(filename, progress_fn, task, job_name):
+    """Continue as long as we detect progress.
+    Ignore the .exit file. Instead, watch for changes to progress_fn.
```
Comment (pb-cdunn) [inline src/py/mains/run.py]: Amendment. This version is just like the old waiter, except that it *also* quits if there is no progress.

File: src/py/mains/run.py [hunk 2]
```diff
@@ -111,31 +120,141 @@ def wait_for_file(filename, task, job_name = ""):
...
+                break
+            else:
+                continue
+        last_progress = latest_progress
+        last_update = latest_update
+
+# Compare nap to MAX_UPDATE_DELTA_S.
+# nap should be less!
+WRAPPER_SCRIPT = """
+progress() {
```
Comment (pb-cdunn) [inline src/py/mains/run.py]: I also tried a version which creates a separate progress script to run in the background. That makes it slightly easier to avoid dangling processes. However, if someday runs `kill -9 PID`, there is nothing we can do anyway. Whoever is killing out submitted jobs (if that is what's happening) is responsible for running `kill -9 -PID`, with a negative PID to kill the entire process group.

**Comment by pb-cdunn** [comment]
Btw, I was trapping only EXIT, which includes so-called "soft signals" like SIGHUP, but not SIGINT nor SIGTERM. So adding those to the bash 'trap' might be simpler solution. Also, I think we might need to run sub-procs in the background and 'wait' on them, so that signals can be handled. That's definitely true for sleep, but maybe for others too. I'll push a small change for those cases.

However, this PR will work even for SIGKILL, since it does not rely on 'trap' at all.

**Comment by pb-jchin** [comment]
Just a comment, the assumption that will work well is that the file system is more reliable (or more robust or easier to configure) for communication than going through network socket. This could be true in many of our current deployment environment.  It is a compromise in my opinion.  Ideally, going through network to get heart beats seems more elegant but practically it might have a lot of problems.  @pb-cdunn, I actually think it is good. I know you will have better ideas eventually.

**Comment by pbjd** [comment]
Assume NFS here, which doesn't suffer from the 'ghost-write' issues Jason found with lustre.

**Comment by pbjd** [comment]
We also need a method to reproduce the problem so we can test the fix.

**Comment by pb-cdunn** [comment]
Well, I've tested simply killing a process. And I've also tested behavior by adding a sleep to a job and reducing the delta to zero. It behaves as expected. (I'll write much better tests when the job-distributor is architected as a separate process. It will also be easier to add CPU time tracking, for benchmarking.)

I can think of some pitfalls. E.g. if whatever is killing the process fails to kill the process-group, then we'll end up with a bunch of dangling "sleeps". Not a big deal though, and that would be a nice clue about the root cause.

Also, it's possible that my change in f4a87add24d6e9c00349da7945bcda4b0b99b260 already fixed the problem.... Well, it turns out that `trap EXIT` already traps SIGTERM, but not SIGINT. And of course not SIGKILL. SIGINT seems unlikely, and SIGKILL cannot be "trapped". So I kinda doubt that change solved anything. Perfectly safe though.

I'm not sure we'll ever know the exact root cause, but maybe we can avoid the problem now.
